printer.go

```diff
@@ -35,113 +47,58 @@ func NewPrinter() *Printer {
 		ValueInterval:   500 * time.Millisecond,
 	}
 }
-
-const nPrinterOut = len(printerHeadings)
-
-var (
-	printerHeadings = [...]string{
-		"Iter",
-		"FunEval",
-		"Obj",
-		"GradNorm",
-	}
-)
-
-func (p *Printer) Init(f *FunctionInfo) error {
-	p.printGrad = f.IsFunctionGradient || f.IsGradient
-
-	p.lastHeading = p.HeadingInterval + 1          // So the headings are printed the first time
-	p.lastValue = time.Now().Add(-p.ValueInterval) // So the values are printed the first time
+func (p *Printer) Init() error {
+	p.lastHeading = p.HeadingInterval
+	p.lastValue = time.Now().Add(-p.ValueInterval)
 	return nil
 }
 
 func (p *Printer) Record(loc *Location, _ EvaluationType, iter IterationType, stats *Stats) error {
-	// Only print on major and initial iterations, or if the iteration is over.
 	if iter != MajorIteration && iter != InitIteration && iter != PostIteration {
 		return nil
 	}
 
-	var nPrint int
-	if p.printGrad {
-		nPrint = nPrinterOut
-	} else {
-		nPrint = nPrinterOut - 1
+	// Print values always on PostIteration or when ValueInterval has elapsed.
+	printValues := time.Since(p.lastValue) > p.ValueInterval || iter == PostIteration
+	if !printValues {
+		// Return early if not printing anything.
+		return nil
 	}
 
-	// Make the value strings
-	var valueStrings [nPrinterOut]string
-	valueStrings[0] = strconv.Itoa(stats.MajorIterations)
-	valueStrings[1] = strconv.Itoa(stats.FuncEvaluations)
-	valueStrings[2] = fmt.Sprintf("%g", loc.F)
-	if p.printGrad {
-		norm := floats.Norm(loc.Gradient, math.Inf(1))
-		valueStrings[3] = fmt.Sprintf("%g", norm)
+	// Print heading when HeadingInterval lines have been printed, but never on PostIteration.
+	printHeading := p.lastHeading >= p.HeadingInterval && iter != PostIteration
+	if printHeading {
+		p.lastHeading = 1
+	} else {
+		p.lastHeading++
 	}
 
-	var maxLengths [nPrinterOut]int
-
-	for i := 0; i < nPrint; i++ {
-		v := len(printerHeadings[i])
-		v2 := len(valueStrings[i])
-		if v > v2 {
-			maxLengths[i] = v
-			continue
+	if loc.Gradient == nil {
```

Comment:
This style isn't scalable. When constraints come, we'll also (maybe) have constraint violation and constraint gradient. It's preferable to start the string with the set of standard outputs (runtime, iter, etc.), and then append.